x-pack/plugins/apm/public/components/app/settings/agent_explorer/agent_list/index.tsx

```diff
@@ -0,0 +1,219 @@
+                <span className="eui-textTruncate">{serviceName}</span>
+              </EuiFlexItem>
+            </EuiFlexGroup>
+          }
+        />
+      ),
+    },
+    {
+      field: AgentExplorerFieldName.Environments,
+      name: i18n.translate(
+        'xpack.apm.agentExplorerTable.environmentColumnLabel',
+        {
+          defaultMessage: 'Environment',
+        }
+      ),
+      width: `${unit * 16}px`,
+      sortable: true,
+      render: (_, { environments }) => (
+        <EnvironmentBadge environments={environments} />
+      ),
+    },
+    {
+      field: AgentExplorerFieldName.Instances,
+      name: i18n.translate(
+        'xpack.apm.agentExplorerTable.instancesColumnLabel',
+        {
+          defaultMessage: 'Instances',
+        }
+      ),
+      width: `${unit * 8}px`,
+      sortable: true,
+    },
+    {
+      field: AgentExplorerFieldName.AgentName,
+      width: `${unit * 12}px`,
+      name: i18n.translate(
+        'xpack.apm.agentExplorerTable.agentNameColumnLabel',
+        { defaultMessage: 'Agent Name' }
+      ),
+      sortable: true,
+    },
+    {
+      field: AgentExplorerFieldName.AgentVersion,
+      name: i18n.translate(
+        'xpack.apm.agentExplorerTable.agentVersionColumnLabel',
+        { defaultMessage: 'Agent Version' }
+      ),
+      width: `${unit * 8}px`,
+      render: (_, { agentVersion }) => (
+        <ItemsBadge
+          items={agentVersion}
+          multipleItemsMessage={i18n.translate(
+            'xpack.apm.agentExplorerTable.agentVersionColumnLabel.multipleVersions',
+            {
+              values: { versionsCount: agentVersion.length },
+              defaultMessage:
+                '{versionsCount, plural, one {1 version} other {# versions}}',
+            }
+          )}
+        />
+      ),
+    },
+    {
+      field: AgentExplorerFieldName.AgentDocsPageUrl,
+      name: i18n.translate(
+        'xpack.apm.agentExplorerTable.agentDocsColumnLabel',
+        { defaultMessage: 'Agent Docs' }
+      ),
+      width: `${unit * 8}px`,
+      render: (_, { agentName, agentDocsPageUrl }) => (
+        <EuiToolTip content={`${agentName} agent docs`}>
+          <AgentExplorerDocsLink
+            agentName={agentName as AgentName}
+            repositoryUrl={agentDocsPageUrl}
+          />
+        </EuiToolTip>
+      ),
+    },
+  ];
+}
+
+interface Props {
+  items: AgentExplorerItem[];
+  noItemsMessage: React.ReactNode;
+  isLoading: boolean;
+}
+
+export function AgentList({ items, noItemsMessage, isLoading }: Props) {
+  const [selectedAgent, setSelectedAgent] = useState<AgentExplorerItem>();
+  const [showFlyout, setShowFlyout] = useState(false);
+
+  useEffect(() => {
+    setShowFlyout(!!selectedAgent);
+  }, [selectedAgent]);
+
+  const onAgentSelected = (agent: AgentExplorerItem) => {
+    setSelectedAgent(agent);
+  };
+
+  const onCloseFlyout = () => {
+    setShowFlyout(false);
+    setSelectedAgent(undefined);
```

Comment:
Can you not handle this by just setting `selectedAgent`, then also use that to determine whether the flyout should be shown? That means you also don't have to cast `selectedAgent as AgentExplorerItem`, TS will refine it for you